### ISSUE TITLE
Local package imports do not auto complete unless I've already partially typed the folder it resides in

### ISSUE THREAD

**Issue by exts** [flutter/flutter-intellij#3506]
## Steps to Reproduce

Video Reproduction: https://streamable.com/93vlm

1. Create a file in a sub folder inside your `lib/` folder. (example/my_test_class.dart)
1.1 Add a `MyTestClass` class inside the file and save
2. Open your main.dart file
3. Partially type `import` so you can hit enter and the cursor brings you inside the single quotations.
4. Start typing the filename 'my_test_class' and notice the local package file won't auto complete for you.
5. type the folder path instead, hit enter to auto complete the folder path THEN start typing the filename and notice the auto complete works.

Now if you just type `var blah = MyTestClass();` inside of your main then alt+enter, it'll properly import, but if I haven't written a line of code this should properly allow me to auto complete local package filenames in sub folders. It works for dart libraries and 3rd party packages without any issue. **Problem only resides w/ your local application sub folder files**

## Version info

```
[√] Flutter (Channel stable, v1.5.4-hotfix.2, on Microsoft Windows [Version 10.0.17134.765], locale en-US)
    • Flutter version 1.5.4-hotfix.2 at D:\flutter
    • Framework revision 7a4c33425d (3 weeks ago), 2019-04-29 11:05:24 -0700
    • Engine revision 52c7a1e849
    • Dart version 2.3.0 (build 2.3.0-dev.0.5 a1668566e5)

[√] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    • Android SDK at D:\Android\sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = D:\Android\sdk
    • Java binary at: D:\Program Files\JetBrains\apps\AndroidStudio\ch-0\183.5522156\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)
    • All Android licenses accepted.

[√] Android Studio (version 3.4)
    • Android Studio at D:\Program Files\JetBrains\apps\AndroidStudio\ch-0\183.5522156
    • Flutter plugin version 35.3.1
    • Dart plugin version 183.6270
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1343-b01)

[√] IntelliJ IDEA Community Edition (version 2019.1)
    • IntelliJ at D:\Program Files\JetBrains\apps\IDEA-C\ch-0\191.7141.44
    • Flutter plugin version 34.0.4
    • Dart plugin version 191.6183.87

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.
```

**Note:** This was an issue I had with visual studio code at one point too which is why I switched to android studio. 🤔

**Comment by DanTup** [comment]
I *think* this might be by design:

https://github.com/dart-lang/sdk/blob/c4e8cf469a0ff7c19fab0578b496b3593b17a1d5/pkg/analysis_server/lib/src/services/completion/dart/uri_contributor.dart#L184-L198

I think it only looks at the immediate children of the current location, to avoid scanning a potentially large directory tree. I'm not sure about IntelliJ/AS, but in VS Code we re-trigger completion each time you select a folder to make it easier to jump down the tree.

**Comment by exts** [comment]
Seems kind of tedious if every core and 3rd party library supports it, but my custom code doesn't :/

Also wasn't the jetbrains language framework built to index stuff like this as a core feature to their ide's? I wouldn't even mind if it was an opt in feature,

**Comment by DanTup** [comment]
I don't know for certain, I'm just guessing from looking at the code.

> Also wasn't the jetbrains language framework built to index stuff like this as a core feature to their ide's?

Someone more familiar with this project would be better answering this, but my understanding is that IntelliJ used the same language server the other editors (like VS Code - the one I work on) do. I don't know how much it supplements the servers results with its own though.

@bwilkerson do you know if this is a deliberate choice? If the server has a list of files in memory, maybe it wouldn't be too expensive to build a full list?

**Comment by bwilkerson** [comment]
> Also wasn't the jetbrains language framework built to index stuff like this ...

Yes, but the analysis server does not provide any way for clients to pass that kind of information to it, partially because it isn't hard for server to compute that information for itself.

> I think it only looks at the immediate children of the current location ...

Yes, the current implementation will only suggest direct children of the preceding directory.

> ... do you know if this is a deliberate choice?

Yes it was deliberate. Historically server has been designed to complete a single identifier rather than longer chains, and completing a single component in a path was seen as being analogous. That said, we've started thinking about revisiting that design decision and producing longer sequences.

The open question is how many suggestions to provide. If you have a large number of files under `lib`, then the completion menu could be overwhelmed by the number of suggestions, whereas the number of direct children is typically fairly small, making it much easier for the user to find the component they're looking for. I'm hoping that we can run some usability studies at some point to understand better when having more suggestions is better and when it's worse.

@InMatrix  @lambdabaa (Mostly as another data point in ongoing discussions.)

**Comment by pq** [comment]
@bwilkerson: should we move this to the SDK for ongoing discussion (since it's an analysis server issue)?
